cs/src/Management/TunnelManagementClient.cs

```diff
@@ -549,12 +551,24 @@ private string UserLimitsPath
 
             options?.SetRequestOptions(request);
 
-            var response = await this.httpClient.SendAsync(request, cancellation);
-            var result = await ConvertResponseAsync<TResult>(
-                method,
-                response,
-                cancellation);
-            return result;
+            try
+            {
+                var response = await this.httpClient.SendAsync(request, cancellation);
+                var result = await ConvertResponseAsync<TResult>(
+                method,
+                response,
+                cancellation);
+                return result;
+            }
+            catch (HttpRequestException ex) when (ex.InnerException is AuthenticationException auex)
+            {
+                throw new HttpRequestException($"Error: Tunnel service HTTPS certificate is invalid. This may" +
+                    $" be caused by the use of a firewall intercepting the connection.", auex);
+            }
+            catch (Exception ex)

```

Comment:
Is this catch clause needed? The exception may already be of type `HttpRequestException`, and in that case re-wrapping it loses some details. And otherwise I'm not sure it is good to change the type of exception.

Reply:
Exception messages are read-only. The only way that I found to be able to alter the exception message was to create a new exception and use the older exception as the inner exception. This is the same exception with the same inner exception, so minimal details should be lost

Reply:
My comment above is referring to these lines:
```C#
catch (Exception ex)
{
    throw new HttpRequestException(ex.Message);
}
```

This re-throw is not changing the exception message.

When we get to this `catch` block, either the exception is not of type `HttpRequestException` or it that type but without the expected inner exception type. In either case it is not certificate-related so there is no need to handle it. Wrapping it in `HttpRequestException` does not provide any benefit.

Reply:
Good catch, initially with the comment I thought you were referring to the other catch block.

This makes sense, I went ahead and removed this catch block entirely since the exception would just have been rethrown anyways.